lib/parachain/types.go

```diff
@@ -0,0 +1,407 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package parachain
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// TODO: resolve these
+// https://github.com/ChainSafe/gossamer/pull/3241#discussion_r1187883189
+// https://github.com/ChainSafe/gossamer/pull/3297#discussion_r1214740051
+
+// ValidatorIndex Index of the validator. Used as a lightweight replacement of the `ValidatorId` when appropriate
+type ValidatorIndex uint32
+
+// ValidatorID The public key of a validator.
+type ValidatorID [sr25519.PublicKeyLength]byte
+
+// BlockNumber The block number type.
+type BlockNumber uint32
+
+// GroupRotationInfo A helper data-type for tracking validator-group rotations.
+type GroupRotationInfo struct {
+	// SessionStartBlock is the block number at which the session started
+	SessionStartBlock BlockNumber `scale:"1"`
+	// GroupRotationFrequency indicates how often groups rotate. 0 means never.
+	GroupRotationFrequency BlockNumber `scale:"2"`
+	// Now indicates the current block number.
+	Now BlockNumber `scale:"3"`
+}
+
+// ValidatorGroups represents the validator groups
+type ValidatorGroups struct {
+	// Validators is an array of validator set Ids
+	Validators [][]ValidatorIndex `scale:"1"`
+	// GroupRotationInfo is the group rotation info
+	GroupRotationInfo GroupRotationInfo `scale:"2"`
+}
+
+// ParaID Unique identifier of a parachain.
+type ParaID uint32
+
+// GroupIndex The unique (during session) index of a validator group.
+type GroupIndex uint32
+
+// CollatorID represents the public key of a collator
+type CollatorID [sr25519.PublicKeyLength]byte
+
+// CollatorSignature is the signature on a candidate's block data signed by a collator.
+type CollatorSignature [sr25519.SignatureLength]byte
+
+// ValidationCodeHash is the blake2-256 hash of the validation code bytes.
+type ValidationCodeHash common.Hash
+
+// CandidateDescriptor is a unique descriptor of the candidate receipt.
+type CandidateDescriptor struct {
+	// The ID of the para this is a candidate for.
+	ParaID uint32 `scale:"1"`
+
+	// RelayParent is the hash of the relay-chain block this should be executed in
+	// the context of.
+	// NOTE: the fact that the hash includes this value means that code depends
+	// on this for deduplication. Removing this field is likely to break things.
```

Comment:
this looks some comment polkadots folks added for their reference. we don't need this.

Reply:
Yes, I've been using it for my reference. I'll be cleaning these up before merging!